### PR TITLE
Add quiet limit order short circuit. Refactor tests to use pkg version

### PR DESCRIPTION
**Pull request by jarry-xiao**
The proposed contract changes here are extremely small. The goal is to prevent limit order instructions from rolling back the full transaction if the user has insufficient funds.

The new test checks this logic against multiple different inputs.

Changes:
- Create an `amount` function for the token checker. This is pulled directly from Anchor: https://github.com/coral-xyz/anchor/blob/master/spl/src/token.rs#L534
- Create logic in the new order processor to quietly exit the function in the case of insufficient funds:
    1) Check that the user specified that the tx should fail silently
    2) Check the available quote lots and base lots (free + token balance)
    3) For either bid or ask, make sure that the trader has sufficient available balance to perform the new operation
    4) If the funds are insufficient, exit
- Update the test file to use a far more recent version of phoenix-sdk (this is the majority of the changes in this PR). Note that this doesn't affect the contract at all
- Write a new test case to validate that the short circuiting logic works as intended
- Add a new field to OrderPacket to enable the silent failing case.
- Update the deserializer to maintain backwards compatibility

### CODE REVIEW

File: tests/test_phoenix.rs [hunk 1]
```diff
@@ -3440,3 +3560,270 @@ async fn test_phoenix_basic_with_raw_base_unit_adjustment() {
         first_cross_size * base_lot_size + second_cross_size * base_lot_size
     );
 }
+
+#[tokio::test]
+async fn test_phoenix_place_order_quiet_failure() {
```
Comment (jarry-xiao) [inline tests/test_phoenix.rs:3565]: This is the only test that needs review

File: src/program/processor/new_order.rs [hunk 2]
```diff
@@ -325,6 +325,58 @@ fn process_new_order<'a, 'info>(
...
+            let quote_lots_free = market
+                .get_trader_state_from_index(trader_index)
+                .quote_lots_free;
+            let base_lots_free = market
+                .get_trader_state_from_index(trader_index)
+                .base_lots_free;
+            let (quote_lots_available, base_lots_available) = match vault_context.as_ref() {
+                None => (quote_lots_free, base_lots_free),
+                Some(vc) => {
+                    let quote_account_atoms = vc.quote_account.amount().map(QuoteAtoms::new)?;
```
Comment (throwbackjams) [inline src/program/processor/new_order.rs:343]: Just to make sure, this `vault_context.quote_account` is the user's quote account?

File: src/program/processor/new_order.rs [hunk 3]
```diff
@@ -325,6 +325,58 @@ fn process_new_order<'a, 'info>(
...
+                            order_packet.num_base_lots()
+                        );
+                        return Ok(());
+                    }
+                }
+                Side::Bid => {
+                    let quote_lots_required = order_packet.get_price_in_ticks()
+                        * market.get_tick_size()
+                        * order_packet.num_base_lots()
+                        / market.get_base_lots_per_base_unit();
```
Comment (throwbackjams) [inline src/program/processor/new_order.rs:366]: Do we not have an existing conversion function for this?

Comment (jarry-xiao) [reply src/program/processor/new_order.rs:366]: I guess you could also put it on `market`, but I don't want to touch the matching engine (this conversion is used there too)

File: src/program/processor/new_order.rs [hunk 4]
```diff
@@ -325,6 +325,58 @@ fn process_new_order<'a, 'info>(
...
+        if order_packet.fail_silently_on_insufficient_funds() {
+            let trader_index = market
+                .get_trader_index(trader.key)
+                .ok_or(PhoenixError::TraderNotFound)?;
+            let quote_lots_free = market
+                .get_trader_state_from_index(trader_index)
+                .quote_lots_free;
+            let base_lots_free = market
+                .get_trader_state_from_index(trader_index)
+                .base_lots_free;
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:339]: nit: can we fetch these in one go?

File: src/program/processor/new_order.rs [hunk 5]
```diff
@@ -325,6 +325,58 @@ fn process_new_order<'a, 'info>(
...
+            let (quote_lots_available, base_lots_available) = match vault_context.as_ref() {
+                None => (quote_lots_free, base_lots_free),
+                Some(ctx) => {
+                    let quote_account_atoms = ctx.quote_account.amount().map(QuoteAtoms::new)?;
+                    let base_account_atoms = ctx.base_account.amount().map(BaseAtoms::new)?;
+                    (
+                        quote_lots_free + quote_account_atoms.unchecked_div(quote_lot_size),
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:346]: `quote_account_atoms / QuoteAtomsPerQuoteLot::new(quote_lot_size)`

File: src/program/processor/new_order.rs (hunk 4 above)
Comment (eugene-chen) [reply src/program/processor/new_order.rs:339]: alternately just put the whole thing inside the Side match?

File: src/program/processor/new_order.rs [hunk 6]
```diff
@@ -325,6 +325,58 @@ fn process_new_order<'a, 'info>(
...
+                        base_lots_free + base_account_atoms.unchecked_div(base_lot_size),
+                    )
+                }
+            };
+            match side {
+                Side::Ask => {
+                    if base_lots_available < order_packet.num_base_lots() {
+                        phoenix_log!(
+                            "Insufficient funds to place order: {} base lots available, {} base lots requested",
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:355]: requested --> required, here and below

File: src/program/validation/checkers/token_checkers.rs [hunk 7]
```diff
@@ -97,6 +97,15 @@ impl<'a, 'info> TokenAccountInfo<'a, 'info> {
     }
 }
 
+impl<'a, 'info> TokenAccountInfo<'a, 'info> {
+    pub fn amount(&self) -> Result<u64, ProgramError> {
```
Comment (eugene-chen) [inline src/program/validation/checkers/token_checkers.rs:101]: is this copied from somewhere?

if so, source would be nice
if not, struct def in here would be nice

File: src/state/order_schema/order_packet.rs [hunk 8]
```diff
@@ -132,6 +138,10 @@ pub enum OrderPacket {
 
         /// If this is set, the order will be invalid after the specified unix timestamp
         last_valid_unix_timestamp_in_seconds: Option<u64>,
+
+        /// The PostOnly and Limit variants have a flag at this position to fail silently on insufficient funds. This
+        /// padding field is used to help with uniformly deserializing the enum.
+        _padding: u8,
```
Comment (eugene-chen) [inline src/state/order_schema/order_packet.rs:144]: i don't really like this

File: src/state/order_schema/order_packet.rs [hunk 9]
```diff
@@ -644,26 +674,34 @@ impl OrderPacket {
...
 
 pub fn decode_order_packet(bytes: &[u8]) -> Option<OrderPacket> {
-    let order_packet = match OrderPacket::try_from_slice(bytes) {
-        Ok(order_packet) => order_packet,
+    // The optional fields at the end of the order packet are not required in the raw input data.
+    let additional_fields = &[
+        0_u8, /* last_valid_slot */
+        0_u8, /* last_valid_unix_timestamp_in_seconds */
+        0_u8, /* fail_silently_on_insufficient_funds */
+    ];
```
Comment (eugene-chen) [inline src/state/order_schema/order_packet.rs:682]: here, are we assuming that any new fields that will get added in the future will get added to all OrderPackets? are we ok with that?

Comment (eugene-chen) [reply src/state/order_schema/order_packet.rs:682]: and technically shouldn't these be reversed in order?

File: src/state/order_schema/order_packet.rs [hunk 10]
```diff
@@ -739,7 +778,9 @@ fn test_decode_order_packet() {
         };
         let bytes = packet.try_to_vec().unwrap();
         let decoded_normal = decode_order_packet(&bytes).unwrap();
-        let decoded_inferred = decode_order_packet(&bytes[..bytes.len() - 2]).unwrap();
+        let num_stripped_bytes = if rng.gen::<f64>() > 0.5 { 1 } else { 3 };
+        let decoded_inferred =
+            decode_order_packet(&bytes[..bytes.len() - num_stripped_bytes]).unwrap();
```
Comment (eugene-chen) [inline src/state/order_schema/order_packet.rs:783]: we should test both cases deterministically here and below

File: tests/test_phoenix.rs [hunk 11]
```diff
@@ -3440,3 +3561,298 @@ async fn test_phoenix_basic_with_raw_base_unit_adjustment() {
...
+        .await
+        .unwrap();
+
+    let market_end = sdk.get_market_orderbook(market).await.unwrap();
+    assert_eq!(
+        market_start.asks.len(),
+        market_end.asks.len(),
+        "Order should have failed silently"
+    );
+
```
Comment (eugene-chen) [inline tests/test_phoenix.rs:3711]: here, let's send another order with fail_silently set to false, and check that tx fails

File: tests/test_phoenix.rs [hunk 12]
```diff
@@ -3440,3 +3561,298 @@ async fn test_phoenix_basic_with_raw_base_unit_adjustment() {
...
+        .unwrap();
+
+    let market_end = sdk.get_market_orderbook(market).await.unwrap();
+
+    assert_eq!(
+        market_start.bids.len(),
+        market_end.bids.len(),
+        "Order should have failed silently"
+    );
+
```
Comment (eugene-chen) [inline tests/test_phoenix.rs:3843]: same thing here

**Review by eugene-chen — COMMENTED** [review]
Should we find a way to support some form of this field for `PlaceMultiple`? It seems inconsistent to only allow it for a single `PlaceOrder` since we should generally expect in the long run that `PlaceMultiple` is used for placing a bunch of orders

File: src/program/processor/new_order.rs (hunk 4 above)
Comment (jarry-xiao) [reply src/program/processor/new_order.rs:339]: yeah, was a little lazy here, can make the code a bit prettier

File: src/program/validation/checkers/token_checkers.rs (hunk 7 above)
Comment (jarry-xiao) [reply src/program/validation/checkers/token_checkers.rs:101]: read PR description: copied from Anchor

File: src/state/order_schema/order_packet.rs (hunk 8 above)
Comment (jarry-xiao) [reply src/state/order_schema/order_packet.rs:144]: I think we might be able to remove it

File: src/state/order_schema/order_packet.rs (hunk 10 above)
Comment (jarry-xiao) [reply src/state/order_schema/order_packet.rs:783]: oh i see what you mean. will do

File: src/program/processor/new_order.rs (hunk 5 above)
Comment (jarry-xiao) [reply src/program/processor/new_order.rs:346]: also quote_lot_size is already typed

File: src/program/processor/new_order.rs [hunk 13]
```diff
@@ -62,7 +108,11 @@ impl MultipleOrderPacket {
             bids,
             asks,
             client_order_id,
-            reject_post_only,
+            failed_multiple_limit_order_behavior: if reject_post_only {
+                FailedMultipleLimitOrderBehavior::RejectPostOnly
+            } else {
+                FailedMultipleLimitOrderBehavior::AmendOnCross
+            },
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:115]: should we change the method signature to include fail_silently_if_insufficient_funds?

File: src/program/processor/new_order.rs [hunk 14]
```diff
@@ -552,10 +627,22 @@ fn process_multiple_new_orders<'a, 'info>(
...
+                    // Decrement the available funds by the amount that was deposited after each iteration
+                    // This should never underflow, but if it does, the program will panic and the transaction will fail
+                    quote_lots_available -=
+                        quote_lots_deposited + matching_engine_response.num_free_quote_lots_used;
+                    base_lots_available -=
+                        base_lots_deposited + matching_engine_response.num_free_base_lots_used;
+                }
+
+                quote_lots_to_deposit += quote_lots_deposited;
+                base_lots_to_deposit += base_lots_deposited
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:645]: semicolon?

File: src/program/processor/new_order.rs [hunk 15]
```diff
@@ -611,3 +698,70 @@ fn process_multiple_new_orders<'a, 'info>(
...
+                base_lots_free + base_account_atoms.unchecked_div(base_lot_size),
+                quote_lots_free + quote_account_atoms.unchecked_div(quote_lot_size),
+            )
+        }
+    };
+    Ok((base_lots_available, quote_lots_available))
+}
+
+fn order_packet_has_sufficient_funds<'a>(
+    market_wrapper: &MarketWrapperMut<'a, Pubkey, FIFOOrderId, FIFORestingOrder, OrderPacket>,
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:734]: this and above would be nice if market_wrapper could be marked as not mut where you can still pass in a mut one if you really want to, cause we're not mutating it

File: src/program/processor/new_order.rs [hunk 16]
```diff
@@ -22,17 +25,60 @@ use solana_program::{
...
+    /// Crossing orders will be amended to the closest non-crossing price.
+    FailSilentlyAndAmend,
+
+    /// Orders will be skipped if the user has insufficient funds.
+    /// Crossing orders will be skipped.
+    FailSilentlyAndRejectCrossingOrders,
+}
+
+impl FailedMultipleLimitOrderBehavior {
+    pub fn should_reject_post_only(&self) -> bool {
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:47]: and then this guy is `should_reject_crossing_orders`

File: src/program/processor/new_order.rs [hunk 17]
```diff
@@ -324,8 +388,28 @@ fn process_new_order<'a, 'info>(
         let clock = Clock::get()?;
         let mut get_clock_fn = || (clock.slot, clock.unix_timestamp as u64);
         let market_bytes = &mut market_info.try_borrow_mut_data()?[size_of::<MarketHeader>()..];
-        let market = load_with_dispatch_mut(&market_info.size_params, market_bytes)?.inner;
-        let (order_id, matching_engine_response) = market
+        let market_wrapper = load_with_dispatch_mut(&market_info.size_params, market_bytes)?;
+        // If the trader does not have sufficient funds to place the order, return silently without mutating the book.
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:392]: comment: "If the order should fail silently on insufficient funds, and the trader does not have sufficient funds for the order, return silently without modifying the book."

File: src/program/processor/new_order.rs [hunk 18]
```diff
@@ -560,3 +729,70 @@ fn process_multiple_new_orders<'a, 'info>(
...
+        (trader_state.base_lots_free, trader_state.quote_lots_free)
+    };
+    let (base_lots_available, quote_lots_available) = match vault_context.as_ref() {
+        None => (base_lots_free, quote_lots_free),
+        Some(ctx) => {
+            let quote_account_atoms = ctx.quote_account.amount().map(QuoteAtoms::new)?;
+            let base_account_atoms = ctx.base_account.amount().map(BaseAtoms::new)?;
+            (
+                base_lots_free + base_account_atoms.unchecked_div(base_lot_size),
+                quote_lots_free + quote_account_atoms.unchecked_div(quote_lot_size),
```
[the remark this answers is not in the document]
Comment (jarry-xiao) [reply src/program/processor/new_order.rs:757]: normal div would emit a warning log, and i don' t think that's necessary here

[the remark this answers is not in the document]
Comment (eugene-chen) [reply src/program/processor/new_order.rs:725]: i see, i guess that means unchecked is doing two things? 1) throwing away units, 2) not warning on unclean division?

File: src/program/processor/new_order.rs [hunk 19]
```diff
@@ -22,17 +25,55 @@ use solana_program::{
...
+
+    /// If any order crosses the spread or has insufficient funds, the entire transaction will fail.
+    FailOnInsufficientFundsAndFailOnCross,
+
+    /// Orders will be skipped if the user has insufficient funds.
+    /// Crossing orders will be amended to the closest non-crossing price.
+    SkipOnInsufficientFundsAndAmendOnCross,
+
+    /// Orders will be skipped if the user has insufficient funds.
+    /// Crossing orders will be skipped.
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:44]: if any order crosses the book, the entire tx will fail

File: src/program/processor/new_order.rs [hunk 20]
```diff
@@ -324,8 +384,29 @@ fn process_new_order<'a, 'info>(
         let clock = Clock::get()?;
         let mut get_clock_fn = || (clock.slot, clock.unix_timestamp as u64);
         let market_bytes = &mut market_info.try_borrow_mut_data()?[size_of::<MarketHeader>()..];
-        let market = load_with_dispatch_mut(&market_info.size_params, market_bytes)?.inner;
-        let (order_id, matching_engine_response) = market
+        let market_wrapper = load_with_dispatch_mut(&market_info.size_params, market_bytes)?;
+        // If the order should fail silently on insufficient funds, and the trader does not have sufficient funds for the order,
```
Comment (eugene-chen) [inline src/program/processor/new_order.rs:388]: nit: newline somewhere? this comment is too long for one line

**Review by eugene-chen — APPROVED** [review]

**Pull request merged**